tabular/src/autogluon/tabular/models/widedeep_nn/widedeep_nn_model.py

```diff
@@ -0,0 +1,289 @@
+from autogluon.core import metrics
+from autogluon.core.constants import BINARY, REGRESSION, MULTICLASS
+from autogluon.core.models import AbstractModel
+from autogluon.core.utils.exceptions import TimeLimitExceeded
+from autogluon.core.utils.files import make_temp_directory
+from autogluon.core.utils.try_import import try_import_pytorch_widedeep
+from .hyperparameters.parameters import get_param_baseline
+from .hyperparameters.searchspaces import get_default_searchspace
+from .metrics import get_nn_metric, get_objective, get_monitor_metric
+from .preprocessing_utils import MissingFiller, TargetScaler, CategoricalFeaturesFilter
+from .utils import set_seed
+
+logger = logging.getLogger(__name__)
+
+
+class WideDeepNNModel(AbstractModel):
+
+    def __init__(self, path: str = None, name: str = None, problem_type: str = None, eval_metric: Union[str, metrics.Scorer] = None, hyperparameters=None):
+        super().__init__(path, name, problem_type, eval_metric, hyperparameters)
+        self.y_scaler = None
+        self.missing_filler = None
+        self.cont_normalization = None
+
+    # TODO: Leverage sample_weight
+    # TODO: Experiment with text and image data
+    def _fit(self,
+             X,
+             y,
+             X_val=None,
+             y_val=None,
+             sample_weight=None,
+             time_limit=None,
+             num_cpus=None,
+             num_gpus=0,
+             **kwargs):
+        start_time = time.time()
+
+        try_import_pytorch_widedeep()
+
+        from pytorch_widedeep import Trainer
+        from pytorch_widedeep.callbacks import ModelCheckpoint
+        from .callbacks import EarlyStoppingCallbackWithTimeLimit
+        import pytorch_widedeep
+
+        # Deterministic training
+        set_seed(0, True)
+
+        params = self._get_model_params()
+        logger.log(15, f'Fitting with parameters {params}...')
+
+        X_train, X_valid, cont_cols, embed_cols, val_split = self.__prepare_datasets_before_fit(X, y, X_val, y_val)
+
+        nn_metric = get_nn_metric(self.problem_type, self.stopping_metric, self.num_classes)
+        monitor_metric = get_monitor_metric(nn_metric)
+        objective = get_objective(self.problem_type, self.stopping_metric)
+        pred_dim = self.num_classes if self.problem_type == MULTICLASS else 1
+
+        model = self.__construct_wide_deep_model(
+            model_type=self.params['type'],
+            column_idx=self._tab_preprocessor.column_idx,
+            embed_input=None if embed_cols is None else self._tab_preprocessor.cat_embed_input,
+            continuous_cols=cont_cols,
+            pred_dim=pred_dim,
+            **self.params.get('model_args', {})
+        )
+        logger.log(15, model)
+
+        best_epoch = None
+        best_epoch_stop = params.get("best_epoch", None)  # Use best epoch for refit_full.
+        batch_size = self.__get_batch_size(params)
+
+        tab_opt = torch.optim.Adam(model.deeptabular.parameters(), lr=(params['lr']))
```

Comment:
Why `lr=(params['lr'])` instead of `lr=params['lr']`?

Reply:
fixed